packages/authorizer/certora/specs/Authorizer.spec

```
@@ -1,3 +1,386 @@
+
+invariant checkCount(address who, address where)
+    ghostCount[who][where] == getPermissionsLength(who, where);
+
+invariant validPermissionsState(address who, address where)
+    forall bytes4 what . (ghostCount[who][where] == 0 => !ghostAuthorized[who][where][what])
+        && (!ghostAuthorized[who][where][what] => ghostParamsLength[who][where][what] == 0)
+        && (ghostAuthorized[who][where][what] => ghostCount[who][where] > 0);
+
+// TODO: (exists bytes4 what . ghostCount[who][where] > 0 => ghostAuthorized[who][where][what])
+
 use rule sanity;
+
+use rule reentrancyGuard filtered {
+    f -> 
+        !f.isView
+        && f.selector != getSelector(CHANGE_PERMISSIONS())
+        && f.selector != getSelector(INITIALIZE())
+}
+
+rule senderMustBeAuthorizedToAuthorizeOthers(env e, address who, address where, bytes4 what, IAuthorizer.Param[] params)
+    good_description "Only authorized users can authorize other users"
+{
+    bytes4 sigAuthorize = to_bytes4(getSelector(AUTHORIZE()));
+    uint256[] how = helpers.authParams(who, where, what);
+    bool senderCanAuthorize = isAuthorized(e.msg.sender, currentContract, sigAuthorize, how);
+
+    authorize(e, who, where, what, params);
+
+    assert senderCanAuthorize;
+}
+
+rule senderMustBeAuthorizedToUnauthorizeOthers(env e, address who, address where, bytes4 what)
+    good_description "Only authorized users can unauthorize other users"
+{
+    bytes4 sigAuthorize = to_bytes4(getSelector(UNAUTHORIZE()));
+    uint256[] how = helpers.authParams(who, where, what);
+    bool senderCanUnauthorize = isAuthorized(e.msg.sender, currentContract, sigAuthorize, how);
+
+    unauthorize(e, who, where, what);
+
+    assert senderCanUnauthorize;
+}
+
+rule otherFunctionsCannotAuthorize(env e, method f, calldataarg args, address who, address where, bytes4 what)
+    filtered {
+        f -> 
+            !f.isView
+            && f.selector != getSelector(AUTHORIZE())
+            && f.selector != getSelector(CHANGE_PERMISSIONS())
+            && f.selector != getSelector(INITIALIZE())
+    }
+    good_description "Permissions can only be granted by authorize or changePermissions functions"
+{
+    uint256[] how = helpers.authParams(who, where, what);
+    bool authorizedBefore = isAuthorized(who, where, what, how);
+
+    f(e, args);
+
+    bool authorizedAfter = isAuthorized(who, where, what, how);
+
+    assert !authorizedBefore => !authorizedAfter;
+}
+
+rule unauthorizeAfterAuthorizeShouldNotChangeStorage
+    (env e, address who, address where, bytes4 what, IAuthorizer.Param[] params)
+    good_description "Doing authorize and then unauthorize shouldn't change the storage"
+{
+    require params.length == 0 && ghostReentrancyStatus == 1;
+
+    uint256[] how = helpers.authParams(who, where, what);
+    require !isAuthorized(who, where, what, how);
+    require getPermissionParamsLength(who, where, what) == 0; // TODO: write an invariant for this
```

Isn't there an invariant for this above?

```suggestion
    require getPermissionParamsLength(who, where, what) == 0;
```

---

Yes, but the invariant doesn't work yet